common/unsafe/src/main/java/org/apache/spark/unsafe/memory/MemoryBlock.java

```diff
@@ -45,38 +45,159 @@
+      throw new ArrayIndexOutOfBoundsException(
+        "Length " + length + " and offset " + offset + "must be non-negative");
+    }
+    this.obj = obj;
+    this.offset = offset;
     this.length = length;
   }
 
+  protected MemoryBlock() {
+    this(null, 0, 0);
+  }
+
+  public final Object getBaseObject() {
+    return obj;
+  }
+
+  public final long getBaseOffset() {
+    return offset;
+  }
+
+  public void resetObjAndOffset() {
+    this.obj = null;
+    this.offset = 0;
+  }
+
   /**
    * Returns the size of the memory block.
    */
-  public long size() {
+  public final long size() {
     return length;
   }
 
-  /**
-   * Creates a memory block pointing to the memory used by the long array.
-   */
-  public static MemoryBlock fromLongArray(final long[] array) {
-    return new MemoryBlock(array, Platform.LONG_ARRAY_OFFSET, array.length * 8L);
+  public final void setPageNumber(int pageNum) {
+    pageNumber = pageNum;
+  }
+
+  public final int getPageNumber() {
+    return pageNumber;
   }
 
   /**
    * Fills the memory block with the specified byte value.
    */
-  public void fill(byte value) {
+  public final void fill(byte value) {
     Platform.setMemory(obj, offset, length, value);
   }
+
+  /**
+   * Instantiate MemoryBlock for given object type with new offset
+   */
+  public final static MemoryBlock allocateFromObject(Object obj, long offset, long length) {
+    MemoryBlock mb = null;
+    if (obj instanceof byte[]) {
+      byte[] array = (byte[])obj;
+      mb = new ByteArrayMemoryBlock(array, offset, length);
+    } else if (obj instanceof long[]) {
+      long[] array = (long[])obj;
+      mb = new OnHeapMemoryBlock(array, offset, length);
+    } else if (obj == null) {
+      // we assume that to pass null pointer means off-heap
+      mb = new OffHeapMemoryBlock(offset, length);
+    } else {
+      throw new UnsupportedOperationException(
+        "Instantiate MemoryBlock for type " + obj.getClass() + " is not supported now");
+    }
+    return mb;
+  }
+
+  /**
+   * Just instantiate the same type of MemoryBlock with new offset and size. The data is not
+   * copied. If parameters are invalid, an exception is thrown
+   */
+  public abstract MemoryBlock subBlock(long offset, long size);
+
+  protected void checkSubBlockRange(long offset, long size) {
+    if (this.offset + offset < 0 || size < 0) {
+      throw new ArrayIndexOutOfBoundsException(
+        "Size " + size + " and offset " + (this.offset + offset) + " must be non-negative");
+    }
+    if (offset + size > length) {
+      throw new ArrayIndexOutOfBoundsException("The sum of size " + size + " and offset " +
+        offset + " should not be larger than the length " + length + " in the MemoryBlock");
+    }
+  }
+
+  /**
+   * getXXX/putXXX does not ensure guarantee behavior if the offset is invalid. e.g  cause illegal
+   * memory access, throw an exception, or etc.
+   */
+  public abstract int getInt(long offset);
```

Comment:
If we work for `0-based offset`, I think that we may need to update additional many files. This is because many files include hard-coded offset such as `Platform.BYTE_ARRAY_OFFSET`. In particular, the generated code uses the hard-coded offset.
I think that it seems to very hard and buggy work to make it right with introducing `MemoryBlock` at the first version.

Can we prepare two version `getInt()` (0-based offset) and `getIntOffset()` (current semantics)? Then, we first use `getIntOffset()`. In the following JIRAs, we are incrementally change from `getIntOffset()` to `getInt()`? WDYT?


BTW, Here is a list of followup JIRAs:
- SPARK-23779, #20890, [WIP]TaskMemoryManager and UnsafeSorter related classes use MemoryBlock
- SPARK-20871, #23762, [WIP]UTF8StringBuffer uses MemoryBlock
- SPARK-19472, #22246, [WIP]Improve performance of UnsafeRow, UnsafeArrayData, UnsafeMapData, and SpecificUnsafeRowJoiner by using MemoryBlock
- SPARK-20874, #23763, [WIP]OffHeapColumnVector uses MemoryBlock 
These PRs do not use `0-based offset` yet.

Reply:
@cloud-fan After prototyping, I succeeded to make `UTF8String` right in this PR.